go/graph/graph_test.go

```diff
@@ -291,3 +293,53 @@ func TestNeighbors(t *testing.T) {
 		t.Errorf("old graph node has neighbors in new graph: %v", neighbors)
 	}
 }
+
+var allPathExpectation = `2 paths
+path with 3 edges
+ - root -> inter2
+ - inter2 -> inter3
+ - inter3 -> target
+path with 2 edges
+ - root -> inter1
+ - inter1 -> target
+`
+
+func TestAllPathSearch(t *testing.T) {
+	g := New(Directed)
+	nodes := make([]Node, 5)
+	for i := 0; i < 5; i++ {
+		nodes[i] = g.MakeNode()
+	}
+
+	// 0 = root, 1 = inter1, 2 = inter2, 3 = inter3, 4 = target
+
+	// output chain 1: root -> inter2 -> inter3 -> target
+	g.MakeEdge(nodes[0], nodes[2])
+	g.MakeEdge(nodes[2], nodes[3])
+	g.MakeEdge(nodes[3], nodes[4])
+
+	// output chain 2: root -> inter1 -> target
+	g.MakeEdge(nodes[0], nodes[1])
+	g.MakeEdge(nodes[1], nodes[4])
+
+	*nodes[0].Value = "root"
+	*nodes[1].Value = "inter1"
+	*nodes[2].Value = "inter2"
+	*nodes[3].Value = "inter3"
+	*nodes[4].Value = "target"
+
+	paths := g.AllPathSearch(nodes[0], nodes[4])
```

Comment:
This test works, but an easier way to implement a test in the future would likely be to define your expected paths and use `reflect.DeepEqual`.